### ISSUE TITLE
Discounts apportionment issues

### ISSUE THREAD

**Issue by ryanmitchell** [lunarphp/lunar#883]
On the `AmountOff` discount type you are apportioning the discount over the lines:
https://github.com/lunarphp/lunar/blob/a8448755b5c89b2500c1eb11a585d569c392d6b9/packages/core/src/DiscountTypes/AmountOff.php#L80-L89

However on `BuyXGetY` this is not taking place which causes VAT calculation errors. HMRC rules say that discounts like this are a 'pack'... ie you treat the new price as purchasing of all the items, work out price which they would have been without the discount, then apportion the difference across all the items, that way VAT is disbursed correctly.

e.g.

You have item A at £10 and you need to buy 2 to get item B for free which normally costs £5.

Normal price = 2 x A (£10) plus 1 x B (£5) = £25
Discounted price = £20
New price (for Tax) = 2 x A (£8) plus 1 x B (£4) = £20

The customer would still see the price as the full price with a discount line, but for the purposes of taxation this needs correctly calculated.

[Update: I've noticed that FixedAmount discounts also suffer from this issue - in that the discount is spread across the number of items, not weighted according to the price of the items]

**Comment by glennjacobs** [comment]
I'm happy with the calculation proposal. 

In terms of the Cart UI, how would you propose we display it to the customer?

**Comment by wychoong** [comment]
Can this be configurable in someways? There are situation that doesn’t need VAT/HMRC and just simply wants the free item to be $0

Currently can we replace the BuyXgetY/amountOff and substitute with own flavor?

**Comment by ryanmitchell** [comment]
For the end user I don’t want to show the apportioned price at all. But the warehouse system integration needs it to be apportioned 

@wychoong what I’m suggesting won’t affect the existing implantation.

**Comment by glennjacobs** [comment]
I am keen to ensure figures are correct for accounting purposes. I know this is specific to UK VAT, but my understanding is accounting principles are fairly standard.

@ryanmitchell how will you continue to show the item as free in the UI if we're changing the price totals?

**Comment by ryanmitchell** [comment]
I actually wonder if we need to change the approach a little and have a `$line->apportionedTotals` or something like that.

```
$line->apportionedTotals = [
   'subtotal' => x,
   'tax' => y,
];
```

That way you have options:
1. you can output the original values (eg £0 for the discounted item)
2. you have the discountTotal line if you want to show each line original total with a discount amount underneath
3. you have the subTotalDiscounted if you want to show each lines new subtotal inc discount
4. you have the apportioned totals if you want to show the apportionment totals

**Comment by glennjacobs** [comment]
I guess it could work, but it's getting a little complex.

**Comment by glennjacobs** [comment]
Worth checking how Shopify does it, and whether that conforms to HMRC requirements?

**Comment by ryanmitchell** [comment]
They dont seem to handle it until reporting, as they can get away with that sort of thing. I'll try some things and see what I can come up with to keep it as clean as possible, maybe we hide it behind a config variable so those that dont want it dont get it.

**Comment by ryanmitchell** [comment]
Ok, I've pushed some updates to the BuyXGetY branch I was working on: https://github.com/lunarphp/lunar/pull/881/commits/479f2993252a593a8583ce29f60d6148e82993a3

When an item is in a discount it works out what its 'share' of the discount is (based on its price percentage of the overall pack price before discount) and applies this to a new CartLine property `apportionedDiscount`. 

This value would then allow developers to calculate what they want - eg subtotal - apportionedDiscount = taxableSubTotal

I've also added this property to FixedAmount discounts as it was calculating discount differently (based on qty not price).

There also doesn't seem to be any consideration given to multiple discounts applying to the same item - the discount amount gets reset each time, instead of compounded. This will definitely case issues. You'll see in BuyXGetY I've left some comments where this needs to be considered.

Anyway, the intention here is to spark thoughts, not that this is the final solution

**Comment by glennjacobs** [comment]
> There also doesn't seem to be any consideration given to multiple discounts applying to the same item - the discount amount gets reset each time, instead of compounded. This will definitely cause issues. You'll see in BuyXGetY I've left some comments where this needs to be considered.

Should this be a separate issue?

**Comment by ryanmitchell** [comment]
> > There also doesn't seem to be any consideration given to multiple discounts applying to the same item - the discount amount gets reset each time, instead of compounded. This will definitely cause issues. You'll see in BuyXGetY I've left some comments where this needs to be considered.
> 
> Should this be a separate issue?

Pulled it out separately as #885

**Comment by ryanmitchell** [comment]
After some Discord chat its looking like apportionment would need to be handled by the developer, so I've opened up https://github.com/lunarphp/lunar/pull/884 as a starting point for providing the information required for this to be calculated.

**Comment by ryanmitchell** [comment]
Closed by #884
